indexing-service/src/main/java/io/druid/indexing/overlord/PortFinder.java

```diff
@@ -55,30 +58,54 @@ private static boolean canBind(int portNum)
 
   public synchronized int findUnusedPort()
   {
-    int port = chooseNext(startPort);
-    while (!canBind(port)) {
-      port = chooseNext(port + 1);
+    if (candidatePorts != null && !candidatePorts.isEmpty()) {
+      int port = chooseFromCandidates();
+      usedPorts.add(port);
+      return port;
+    } else {
+      int port = chooseNext(startPort);
+      while (!canBind(port)) {
+        port = chooseNext(port + 1);
+      }
+      usedPorts.add(port);
+      return port;
     }
-    usedPorts.add(port);
-    return port;
   }
 
-  public synchronized Pair<Integer, Integer> findTwoConsecutiveUnusedPorts()
+  public synchronized Pair<Integer, Integer> findTwoUnusedPorts()
   {
-    int firstPort = chooseNext(startPort);
-    while (!canBind(firstPort) || !canBind(firstPort + 1)) {
-      firstPort = chooseNext(firstPort + 1);
+    if (candidatePorts != null && !candidatePorts.isEmpty()) {
+      int firstPort = chooseFromCandidates();
+      int secondPort = chooseFromCandidates();
+      usedPorts.add(firstPort);
+      usedPorts.add(secondPort);
+      return new Pair<>(firstPort, secondPort);
+    } else {
+      int firstPort = chooseNext(startPort);
+      while (!canBind(firstPort) || !canBind(firstPort + 1)) {
+        firstPort = chooseNext(firstPort + 1);
+      }
+      usedPorts.add(firstPort);
+      usedPorts.add(firstPort + 1);
+      return new Pair<>(firstPort, firstPort + 1);
     }
-    usedPorts.add(firstPort);
-    usedPorts.add(firstPort + 1);
-    return new Pair<>(firstPort, firstPort + 1);
   }
 
   public synchronized void markPortUnused(int port)
   {
     usedPorts.remove(port);
   }
 
+  private int chooseFromCandidates()
+  {
+    for (int port : candidatePorts) {
+      if (!usedPorts.contains(port) && canBind(port)) {
+        return port;
+      }
+    }
+    throw new ISE("All ports are Used..");
```

Comment:
Instead of throwing an exception, I guess it would be better if `chooseNext()` is used to find a port in case all the candidate ports are unusable.

Reply:
I do not think so. The reason why adding ability to specify list of task ports is that we need to decide ahead of time exactly what ports peon processes are going to use in some containerized environments. So if user specify `druid.indexer.runner.ports`, they shoulde ensure ports are usable. And it may be a illegal thing in users' environments if we use `chooseNext()` to find a port while all the candidate ports are unusable. In my use case, it does a illegal thing and may cause potential port conflicts.

Reply:
IMO, druid nodes should use `druid.indexer.runner.startPort` or `druid.indexer.runner.ports` to allocate task ports rather than using both of them at the same time. What's more, it also will make user confused if him has specified `druid.indexer.runner.ports` but peon processes using ports that not contained in these ports.